File: src/main/java/com/web/SearchWeb/config/ApiResponse.java
package com.web.SearchWeb.config;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class ApiResponse<T> {
    private final boolean success;
    private final T data;
    private final ErrorResponse error; // 여기서 ErrorResponse는 JSON 포장지 역할

    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(true, data, null);
    }

    // Enum을 인자로 받음으로써 타입 안정성 확보
    public static ApiResponse<Void> fail(ErrorCode errorCode) {
        return new ApiResponse<>(
            false,
            null,
            new ErrorResponse(errorCode) // Enum을 생성자로 넘겨서 변환
        );
    }

    @Getter
    public static class ErrorResponse {
        private final String code;
        private final String message;

        // Enum에서 필요한 정보만 추출하여 세팅
        private ErrorResponse(ErrorCode errorCode) {
            this.code = errorCode.getCode();
            this.message = errorCode.getMessage();
        }
    }
}

File: frontend/src/lib/types/apiResponse.ts
// 백엔드 공통 응답 래퍼 (ApiResponse<T>)
// 성공과 실패 케이스를 명확히 분리하여 타입 안정성을 높입니다.
export type ApiResponse<T> =
  | {
      success: true;
      data: T;
      error: null;
    }
  | {
      success: false;
      data: null;
      error: {
        code: string;
        message: string;
      };
    };
